### ISSUE TITLE
derive(PartialEq) shosuld not prevent "field is never read" warnings

### ISSUE THREAD

**Issue by Wilfred** [rust-lang/rust#134588]
### Code

```Rust
#[derive(PartialEq)]
struct MyStruct {
    x: i32,
    y: i32, // no unused field warning, unfortunately
}

struct MyStruct2 {
    x: i32,
    y: i32, // warning today
}

pub fn use_struct() {
    let ms = MyStruct { x: 1, y: 2 };
    let _ = ms.x;

    let ms = MyStruct2 { x: 1, y: 2 };
    let _ = ms.x;
}
```

### Current output

```Shell
warning: field `y` is never read
 --> src/lib.rs:9:5
  |
7 | struct MyStruct2 {
  |        --------- field in this struct
8 |     x: i32,
9 |     y: i32, // warning today
  |     ^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Desired output

```Shell
A warning for both MyStruct and MyStruct2.
```

### Rationale and extra context

This was originally discussed on #84647 and #85200, although the conclusion there was to only exclude Debug and Clone.

However, it's really common to derive PartialEq on a type, especially when writing tests.

This means that adding tests can subtly stop this warning from catching issues. As far as I can see, there isn't a way to opt-in to stricter behaviour with either rustc or clippy here. There's no equivalent of `must_use` for struct fields, for example.

This issue was the root of a nasty bug for me. Would you be open to making this diagnostic fire for more automatically derived traits? 

### Other cases

```Rust

```

### Rust Version

```Shell
Reproduced on rustc 1.83 on the playground.
```

**Comment by ionicmc-rs** [comment]
By the way is just `derive(Trait)` (or at least most derive macros)

**Comment by ionicmc-rs** [comment]
wait so i actually can't reproduce this,  the derived impl is ignored

**Comment by ionicmc-rs** [comment]
Ok ok i reproduced this and it seems only impls of `Debug`, `Clone`, and `Default` are ignored
look at this code:
```rust
#[derive(PartialEq, PartialOrd)]
struct Foo {
    bar: i32,
    baz: i32
}
fn main() {
    let g = Foo { bar: 32, baz: 42 };
    let _ = g.bar;
}
```
in the above code `baz` is used, even though it isnt;
look at the output from `cargo expand` (cargo-expand)
```rust
struct Foo {
    bar: i32,
    baz: i32,
}
#[automatically_derived]
impl ::core::marker::StructuralPartialEq for Foo {}
#[automatically_derived]
impl ::core::cmp::PartialEq for Foo {
    #[inline]
    fn eq(&self, other: &Foo) -> bool {
        self.bar == other.bar && self.baz == other.baz
    }
}
fn main() {
    let g = Foo { bar: 32, baz: 42 };
    let _ = g.bar;
}
```

**Comment by Jules-Bertholet** [comment]
Do we really want to warn here? A field might be read only in `PartialEq` intentionally, e.g. if it serves as a tag of some sort to make instances of the struct unequal that would be equal otherwise. `dead_code` should not have false positives like this IMO

**Comment by smmalis37** [comment]
In addition to `PartialEq` and `PartialOrd` should this issue also track `Hash`?
